builder.md

```diff
@@ -1,57 +1,142 @@
+behaviour of the program.
+
+<!-- TODO: For convenience, we make a manpage and set of tests available. All
+compliant builders MUST fully conform to these two. -->
+
+The CLI interface MUST work without relying on any network connection.
+
+```bash
+base16 TEMPLATES-DIRECTORY SCHEME-FILE ...
+```
+
+`TEMPLATES-DIRETORY` being a directory containing a `config.yaml`, as well as at
+least one `.mustache` template. These are usually named `templates` on all
+base16 template repositories.
+
+`SCHEME-FILE` being one or more `.yaml` scheme file. The main source for these
+is the [base16-schemes](https://github.com/base16-project/base16-schemes)
+repository.
+
+The program MAY offer a single `--debug | -d` option for increasing log
+verbosity. These are considered implementation detail and each author SHOULD
+implement them as they see fit.
+
+### Library
+
+The compliant base16 builder software MAY also expose a software library other
+developers may use to assist developing more complex base16-compatible
+software.
+
+This exposed library, or any internal code, has no required structure or usage.
+
+The author MAY choose how they will expose these functionalities to the caller,
+according to their preferences and SHOULD follow best practices on their
+respective programming languages.
+
+It is RECOMMENDED that builders follow semantic versioning for their library
+interface.
+
+The author MAY implement additional features that are exposed through the
+library, as long as it does not affect the CLI functionality compliance.
+
+## Output and behaviour
+
+**Note**: As the CLI is not intended for usual human usage, all outputted text
+messages are considered implementation detail, so the author MAY output
+whatever they prefer (or no message at all). If needed, scripts using the
+builder SHOULD check for return codes (specified below) instead of messages.
+
+For all templates defined in the template config file (`config.yaml`, inside
+the specified template directory), the builder MUST iterate through all the
+defined schemes and output matching files.
+
+The built filename should look like [output-dir]/base16-[slug][extension],
+where the slug is taken from the scheme filename made lowercase with spaces
+replaced with dashes and both the extension and output-dir are taken from
+`config.yaml`.
+
+The builder MUST check for the (unusual) case where schemes share the same
+slug, in this case the program MUST exit with code `1` and MAY output an error
+message.
+
+If the build fails for whatever other reasons, the program MUST exit with code
+`2` and MAY output an error message.
 
 ## Template Variables
-A builder should provide the following variables to a template file:
+A builder MUST provide the following, and no others, variables to the template
+files it processes:
 
 - `scheme-name` - obtained from the scheme file
 - `scheme-author` - obtained from the scheme file
 - `scheme-slug` - obtained from the scheme filename, as described above
+
```

Comment:
What do you think about having another variable here for dark/light theme? I saw that vscode requires this in their templates and the way to deal with this at the moment would be to have a script that checks for "-light." for each colorscheme name and then add the value to the colorschemes.

Maybe this is something that **should** stay on the template repos side, I just thought I'd bring it up.

Reply:
I think it's a great idea!

If we're asking people to refactor their builders, I don't think it's too much to ask of them. It's actually [something I expose](https://github.com/Misterio77/nix-colors/blob/main/module/colorscheme.nix#L33) with [nix-colors](https://github.com/misterio77/nix-colors).

We could perhaps specify the strategy used to determine the scheme kind  as well, perhaps something simple like: add the three color components together and anything above 382 is dark?